src/Common/EnvironmentProxyConfigurationResolver.cpp

```diff
@@ -0,0 +1,36 @@
+#include "EnvironmentProxyConfigurationResolver.h"
+
+#include <Common/logger_useful.h>
+#include <Poco/URI.h>
+
+namespace DB
+{
+
+static constexpr auto PROXY_HTTP_ENVIRONMENT_VARIABLE = "http_proxy";
+static constexpr auto PROXY_HTTPS_ENVIRONMENT_VARIABLE = "https_proxy";
+
+EnvironmentProxyConfigurationResolver::EnvironmentProxyConfigurationResolver(Protocol protocol_)
+    : protocol(protocol_)
+{}
+
+ProxyConfiguration EnvironmentProxyConfigurationResolver::resolve()
+{
+    bool https = protocol == Protocol::HTTPS;
+
+    if (const auto * proxy_host = std::getenv(https ? PROXY_HTTPS_ENVIRONMENT_VARIABLE : PROXY_HTTP_ENVIRONMENT_VARIABLE)) // NOLINT(concurrency-mt-unsafe)
```

Comment:
It's not that we do not care, it's just that it's the only way to address this?

Reply:
As far as I can tell, there is only one call to [setenv](https://github.com/ClickHouse/ClickHouse/blob/master/src/Daemon/BaseDaemon.cpp#L791) in ClickHouse. I am just not sure it's concurrent, I suppose not. Need to investigate. Pls let me know if you know the answer from the top of your head

Reply:
> As far as I can tell, there is only one call to [setenv](https://github.com/ClickHouse/ClickHouse/blob/master/src/Daemon/BaseDaemon.cpp#L791) in ClickHouse. I am just not sure it's concurrent, I suppose not. Need to investigate. Pls let me know if you know the answer from the top of your head

Yes, it's safe because it's called on an early stage of initialization, there are no other threads yet

Reply:
Added a comment